wandb/sdk/lib/import_hooks.py

```diff
@@ -0,0 +1,251 @@
+"""
+Note: This file is based on
+https://github.com/GrahamDumpleton/wrapt/blob/1.12.1/src/wrapt/importer.py
+(with slight modifications).
+
+This module implements a post import hook mechanism styled after what is
+described in PEP-369. Note that it doesn't cope with modules being reloaded.
+
+"""
+
+import functools
+import importlib  # noqa: F401
+import sys
+import threading
+from typing import Any, Callable, Dict, Optional
+
+
+# modified the following import: from .decorators import synchronized
+def synchronized(lock: "threading.RLock") -> Callable:
+    def decorator(func: Callable) -> Callable:
+        @functools.wraps(func)
+        def new_func(*args: Any, **kwargs: Any) -> Any:
+            with lock:
+                return func(*args, **kwargs)
+
+        return new_func
+
+    return decorator
+
+
+# The dictionary registering any post import hooks to be triggered once
+# the target module has been imported. Once a module has been imported
+# and the hooks fired, the list of hooks recorded against the target
+# module will be truncacted but the list left in the dictionary. This
+# acts as a flag to indicate that the module had already been imported.
+
+_post_import_hooks: Dict = {}
+_post_import_hooks_init: bool = False
+_post_import_hooks_lock = threading.RLock()
+
+# Register a new post import hook for the target module name. This
+# differs from the PEP-369 implementation in that it also allows the
+# hook function to be specified as a string consisting of the name of
+# the callback in the form 'module:function'. This will result in a
+# proxy callback being registered which will defer loading of the
+# specified module containing the callback function until required.
+
+
+def _create_import_hook_from_string(name: str) -> Any:
```

Comment:
the inner katia in me asks "should we remove this unused code path?"

But then i see this as being useful in the future... so. i guess this is good.

Reply:
lol... yeah I assumed it's fine since it's a utility and thought like you said we might use it in the future...

Reply:
sounds good